src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,218 @@
+import { DeviceInputSystem } from '../deviceInputSystem';
+import { Engine } from '../../Engines/engine';
+import { IDisposable } from '../../scene';
+import { DeviceType } from './deviceEnums';
+import { Nullable } from '../../types';
+import { Observable } from '../../Misc/observable';
+
+/**
+ * Class to keep track of devices
+ */
+export class DeviceSourceManager implements IDisposable {
+    // Public Members
+    /**
+     * Observable to be triggered when before a device is connected
+     */
+    public onBeforeDeviceConnectedObservable = new Observable<string>();
+
+    /**
+     * Observable to be triggered when before a device is disconnected
+     */
+    public onBeforeDeviceDisconnectedObservable = new Observable<string>();
+
+    /**
+     * Observable to be triggered when after a device is connected
+     */
+    public onAfterDeviceConnectedObservable = new Observable<string>();
+
+    /**
+     * Observable to be triggered when after a device is disconnected
+     */
+    public onAfterDeviceDisconnectedObservable = new Observable<string>();
+
+    // Private Members
+    private _devices: Array<Array<string>>;
+    private _touchPoints: Array<string>;
+    private _firstDevice: Array<number>;
+    private _deviceInputSystem: DeviceInputSystem;
+
+    /**
+     * Default Constructor
+     * @param engine - engine to pull input element from
+     */
+    constructor(engine: Engine) {
+        const numberOfDeviceTypes = Object.keys(DeviceType).length / 2;
+        this._devices = new Array<Array<string>>(numberOfDeviceTypes);
+        this._firstDevice = new Array<number>(numberOfDeviceTypes);
+        this._deviceInputSystem = new DeviceInputSystem(engine);
+
+        this._deviceInputSystem.onDeviceConnectedObservable.add((deviceName) => {
+            this.onBeforeDeviceConnectedObservable.notifyObservers(deviceName);
+            this._addDevice(deviceName);
+            this.onAfterDeviceConnectedObservable.notifyObservers(deviceName);
+        });
+        this._deviceInputSystem.onDeviceDisconnectedObservable.add((deviceName) => {
+            this.onBeforeDeviceDisconnectedObservable.notifyObservers(deviceName);
+            this._removeDevice(deviceName);
+            this.onAfterDeviceDisconnectedObservable.notifyObservers(deviceName);
+        });
+    }
+
+    // Public Functions
+    /**
+     * Checks for current device input value, given DeviceType, slot, and inputIndex
+     * @param type Enum specifiying device type
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @param inputIndex Index of device input
+     * @returns Current value of input
+     */
+    public getInput(type: DeviceType, inputIndex: number, deviceSlot: number = this._firstDevice[type]): Nullable<number> {

```

Comment:
Correct usage of this function seems not very discoverable or enforceable (e.g. how do you know you are supposed to pass one of the input enum values in for `inputIndex`? You can pass any old number in for `inputIndex` and the code will compile without errors). I think this could be improved with conditional types:
```typescript
type DeviceInputs<T extends DeviceType> =
    T extends DeviceType.Keyboard | DeviceType.GenericDevice ? number :
    T extends DeviceType.Mouse | DeviceType.Touch ? PointerInputs :
    T extends DeviceType.DualShock ? DualShockInputs :
    T extends DeviceType.Xbox ? XboxInputs :
    T extends DeviceType.Switch ? SwitchInputs :
    never;

public getInput<T extends DeviceType>(deviceType: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {
    ...
}
```

This way, when the first parameter (`deviceType`) is known at compile time, then the type of `inputIndex` is both more discoverable and also enforced at compile time. Here is what you see in intellisense when you start typing the function:
![image](https://user-images.githubusercontent.com/5084643/79697392-648d1080-8237-11ea-8c0c-74a96d7cf216.png)
Since `deviceType` has not been specified, `inputIndex` is unconstrained, and intellisense shows exactly the same thing as what it does with your existing implementation. Here is what you see when you've specified `DeviceType.Mouse` for the `deviceType` parameter:
![image](https://user-images.githubusercontent.com/5084643/79697423-a5852500-8237-11ea-9a6f-51bd1094e119.png)
Notice that the `inputIndex` parameter has now been constrained to be of type `PointerInputs`, which is much more discoverable to someone trying to use this API. Further, once `deviceType` has been specified as `DeviceType.Mouse`, then providing anything *other* than `PointerInputs` for the `inputIndex` parameter results in a compile error:
![image](https://user-images.githubusercontent.com/5084643/79697520-26442100-8238-11ea-9551-e5e652794e6f.png)

Reply:
I've implemented the DeviceInputs type to enforce these conditional types